src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDPointSampler.cs

```diff
@@ -0,0 +1,145 @@
+                                                                  StatsDFormatterConstants.ItemTagName
+                                                              };
+
+        private readonly MetricsStatsDOptions _options;
+        private readonly ConcurrentDictionary<string, StatsDSampler> _samplers = new ConcurrentDictionary<string, StatsDSampler>();
+
+        public StatsDPointSampler(MetricsStatsDOptions options) { _options = options; }
+
+        public ConcurrentQueue<StatsDPoint> Points { get; } = new ConcurrentQueue<StatsDPoint>();
+
+        public void Add(
+            string context,
+            string name,
+            string field,
+            object value,
+            MetricTags metricTags,
+            IStatsDMetricStringSerializer serializer,
+            DateTime? timestamp)
+        {
+            if (string.IsNullOrWhiteSpace(name))
+            {
+                throw new ArgumentNullException(nameof(name), "Name could not be null or whitespace");
+            }
+
+            if (string.IsNullOrWhiteSpace(field))
+            {
+                throw new ArgumentNullException(nameof(field), "Field could not be null or whitespace");
+            }
+
+            var tagsDictionary = metricTags.ToDictionary();
+
+            var tags = tagsDictionary.Where(tag => !ExcludeTags.Contains(tag.Key)).ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+
+            tagsDictionary.TryGetValue(AppMetricsConstants.Pack.MetricTagsTypeKey, out var metricType);
+
+            var builder = new StringBuilder();
+
+            if (!string.IsNullOrWhiteSpace(context))
+            {
+                builder.Append(context);
+                builder.Append('.');
+            }
+
+            builder.Append(name.Replace(' ', '_'));
+
+            if (!string.IsNullOrWhiteSpace(metricType))
+            {
+                builder.Append('.');
+                builder.Append(metricType);
+            }
+
+            tagsDictionary.TryGetValue(StatsDFormatterConstants.ItemTagName, out var itemTag);
+            if (!string.IsNullOrWhiteSpace(itemTag))
+            {
+                builder.Append('.');
+                builder.Append(itemTag);
+            }
+
+            builder.Append('.');
+            builder.Append(field);
+
+            var key = builder.ToString();
+            var statsDMetricType = StatsDSyntax.FormatMetricType(metricType);
+
+            if (!StatsDSyntax.CanBeSampled(statsDMetricType))
+            {
+                Points.Enqueue(new StatsDPoint(key, value, statsDMetricType, null, tags, serializer, timestamp));
+                return;
+            }
+
+            tagsDictionary.TryGetValue(StatsDFormatterConstants.SampleRateTagName, out var tagSampleRateStr);
+            var hasTagSampleRate = double.TryParse(tagSampleRateStr, out var tagSampleRate);
+
+            var sampler = _samplers.GetOrAdd(key, _ => new StatsDSampler(_options.DefaultSampleRate));
```

Comment:
@Arkatufus not sure I'm following the sampling concept, could you please provide a bit more context?

Reply:
In StatsD, counter, histogram, and timer metrics can be undersampled by what they call sample rate and it is expressed in percent value. For example, a sample rate of 0.1 means that only 10% of the metric points emitted by the system will be captured and sent to the server; the server then extrapolate the approximate number by scaling the reported numbers.

In this implementation. sample rate can be set as default for all applicable metric type in the `MetricsStatsDOptions` option class or it can be set per metric measurement by adding a "sampleRate" tag to the measurement itself.

In this implementation, the sample filter `StatsDPointSampler` doubles as a StatsD metric name generator. 
1. Generate a unique name for each point. I'm partially following the Datadog implementation for this. The sampled point name is formatted as `[context].[measurement name, spaces replaced with underscore].[app metric metric type].[app metric item key value pair].[field name]`. This name is also used to send the sampled point to the StatsD server.
2. It then checks to see if undersampling can be used for the metric type, if not then it will simply stores it in a write queue  and returns.
3. The generated point name is used as a key to a dictionary of `<string, StatsDSampler>` to store/retrieve a simple sampler instance. I'll explain how the sampler work below.
4. Depending on the point sampler boolean return value, the point is either discarded or stored inside the write queue.
5. All of the points stored inside the queue are then written to the stream when `WriteAsync()` is called.

The point sampler `StatsDSampler` is quite simple. It uses an accumulator with type `double` and adds the `SampleRate` percent value to the accumulator every time the `Sample()` function is called. If the accumulator is greater or equal to one, subtract one from the accumulator and return true, else returns false. The sample point timestamp is used to differentiate each samples with each other, all samples that are captured in the same timestamp are considered as a single measurement and sent to the server.

Reply:
Thanks for the PR and taking the time to provide that context.

DD Ref: https://docs.datadoghq.com/developers/metrics/dogstatsd_metrics_submission/#sample-rates

Reply:
To be honest, it would be great to have a second pair of eyes to check if this implementation is correct or not. I have a nagging suspicion that I have missed something.

Reply:
Can you write a simple spec here for each metric type - set sample rate at 10%, observe the number of attempted data points added (i.e. a reasonably large value) and then test to see that the value reported to the server falls within an acceptable range (using some kind of tolerance / margin) ? That will be more effective than reviewing the code here to find an error case. I think the code looks fine, but a test will be determinate.

Reply:
Question, StatsD sampling is there to reduce network traffic and load on the server i.e. it wouldn't be ideal doing a UDP call on every metric write for client or server.

AppMetrics already aggregates/samples on metric write in memory (on a side note, has some drawbacks), the StatsD reporter implementation here runs on a timer say every x secs. Is it still necessary to apply sampling on top of that?

Reply:
Sort of off topic to this PR, but, it would also be interesting for me, and most probably useful for others, to understand the drivers behind you guys looking at StatsD e.g. why not prom format as most monitoring agents support scraping prom format. Is it just that you want a push based abstraction?

Reply:
> AppMetrics already aggregates/samples on metric write in memory (on a side note, has some drawbacks), the StatsD reporter implementation here runs on a timer say every x secs. Is it still necessary to apply sampling on top of that?

Not really, but I'm trying to implement it as spec. The problem with vanilla StatsD is that its taking "real-time metrics collection" quite literally, ie. you can't really feed StatsD with a timestamped data, unless you make a custom StatsD server that understands how to extract and use the timestamp data to order the incoming data batch, or else it just treats the batch as a single lump of metric data at the time the data is received.

So in order to use StatsD with AppMetrics effectively, the data has to basically try to skip all internal batching and be sent to the StatsD server as soon as it generated as possible.

Reply:
> why not prom format as most monitoring agents support scraping prom format. Is it just that you want a push based abstraction?

Well, we're basing one of our reporting products on AppMetrics and a few of our customers asked if we can implement StatsD as one of the transport protocol.